---
layout: post
title: "What's in your gene list?"
date: "2019-08-06"
slug: "example_content"
description: "How does pathway enrichment actually work."
category:
  - single cell RNAseq
  - genomics
  - statistics
# tags will also be used as html meta keywords.
tags:
  - single cell RNAseq
  - genomics
  - unsupervised
  - RNAseq
show_meta: true
comments: true
mathjax: true
gistembed: true
published: true
noindex: false
nofollow: false
# hide QR code, permalink block while printing.
hide_printmsg: false
# show post summary or full post in RSS feed.
summaryfeed: false
## for twitter summary card with squared image and page description or page excerpt:
# imagesummary: foo.png
## for twitter card with large image:
# imagefeature: "http://img.youtube.com/vi/VEIrQUXm_hY/0.jpg"
## for twitter video card: (active for this page)
#videofeature: "https://www.youtube.com/embed/iG9CE55wbtY"
#imagefeature: "http://img.youtube.com/vi/iG9CE55wbtY/0.jpg"
#videocredit: tedtalks
---

In every genomics experiment there comes a time when you end up with a list of genes. It's usually not  scalable to manually look through these genes especially if you have many different comparisons and samples. So how to best get to the specific signatures and pathways that these genes encode?

<!--more-->

Time for some pathway enrichment or over-representation analysis (ORA)! In the end the hope is to get a table similar to the one shown below.

{:.text-center img}
![Default parameters]({{ site.urlimg }}/media/ORA0.JPG "Default parameters")

This table is part of a recent [Nature protocols article]
(https://www.nature.com/articles/s41596-018-0103-9)
provides comprehensive practical guidance how to perform pathway enrichment.

For the purpose of this blogpost I wanted to generate a tool for a better intuitive understanding of these p-values next to enriched pathways. I made a [Jupyter notebook](https://colab.research.google.com/github/evafast/Jupyter_notebooks/blob/master/190801_intuitiveORA.ipynb) that allows to interactively change each component of the enrichment analysis (size of overlap, size of gene sets, size of gen universe) and visualize how it influences the outcome p-value of the enrichment test (one-tailed Fisher exact test or hypergeometric test).

A few take home points:

### 1) Extra caution for small gene sets

Let's start with the default example in the Jupyter notebook. In this example there are 200 upregulated genes, a GO-term with 50 genes and 10 genes that overlap between the two groups. The total number of genes measured was 10,000. The p-value of the one-tailed Fisher exact test is 4*10^-8 and therefore the overlap can be considered statistically significant.

{:.text-center img}
![Default parameters]({{ site.urlimg }}/media/ORA1.png "Default parameters")

What happens now if there is a smaller overlap of not 10 but 2 genes? Not that surprisingly a smaller overlap of only 2 genes increases the p-value to 0.26 and renders it thus non statistically significant. What if the size of the GO-term is now not 50 genes but smaller for example 10 genes? It could be that there is a specific list of 10 genes that have been associated with a particular disease state or specialized pathway. With a gene set of only 10 genes an overlap of 2 will be become again statistically significant (p = 0.016). Feel free to try this out yourself in the linked [Jupyter notebook](https://colab.research.google.com/github/evafast/Jupyter_notebooks/blob/master/190801_intuitiveORA.ipynb)!

To generalize: When leaving everything else constant the p-value of overlap increases with the size of the GO-term (as well as genes upregulated). This is illustrated in the Figure below.

{:.text-center img}
![Default parameters]({{ site.urlimg }}/media/ORA2.png "Default parameters")

Therefore one needs to be careful with gene-sets and GO-terms that only include a small number of genes. Even an overlap of 1 or 2 genes can give a statistically significant enrichment of that pathway. It has to be carefully evaluated if that enrichment makes biological sense. The review mentioned above actually suggests to exclude all pathways that contain less than 10-15 genes.


### 2) Size of the gene universe matters

Another factor that is frequently overlooked in ORA is the size of the "gene universe". Different experimental methods can vary in the total number of genes they are able to assess. For example droplet based single-cell RNA sequencing experiments are not able to detect low expressed genes. The total gene pool from which potential upregulated gene lists can be sampled are thus reduced. The quantity of the gene universe is usually not readily available in online enrichment tools. For a thorough analysis it's worthwhile to poke around a bit to see what number is being used and how that compares to the specific experiment from which the gene list was generated. To illustrate how gene universe size affects the enrichment result let's go back to the original example. Everything was kept the same (overlap, number of genes up, size of GO-term) but the gene universe was changed from 10,000 genes to 1,000 genes. When sampling from this smaller pool the same overlap is now not statistically significant (p-value: 0.56) anymore.

{:.text-center img}
![Default parameters]({{ site.urlimg }}/media/ORA3.png "Default parameters")

Plotting the relationship of gene universe size versus p-value as in the graph below shows an inverse relationship. The larger the pool of genes that is being assessed the less likely (smaller p-value) it will be that an overlap between two populations is due to random chance.

{:.text-center img}
![Default parameters]({{ site.urlimg }}/media/ORA4.png "Default parameters")

**Some final thoughts:**

For the typical analysis enrichment of many pathways is assessed in which case it is important to do a [multiple test correction](https://www.nature.com/articles/nbt1209-1135) and use the resulting q-value as in the table above as an estimator of significance of overlap. A big limitation of the Fisher-Exact test is that one needs to make a specific cutoff above which genes are included or excluded in the analysis. Sometimes this cutoff can be quite arbitrary and it makes more sense to work with a ranked gene list such as Gene Set Enrichment Analysis or GSEA. I listed some links to a number of online tools for pathway enrichment analysis. Many of these tools use more advanced versions of the traditional Fisher-Exact test to overcome some of its limitations.


[Metascape](http://metascape.org) - also has a good explanation on pathway enrichment on it's [blog] (http://metascape.org/blog/)
[g:Profiler](https://biit.cs.ut.ee/gprofiler/gost)
[GSEA](http://software.broadinstitute.org/gsea/index.jsp)
[David](https://david.ncifcrf.gov/)
[Ingenuity Pathway analysis (IPA)](https://www.qiagenbioinformatics.com/products/ingenuity-pathway-analysis/) - unfortunately not free
